### PR TITLE
MET-28: In measurement family, modal for adding a new unit

### PR DESCRIPTION
**Pull request by tseho**
**Description (for Contributor and Core Developer)**

Screenshot_2020-03-20_17-52-02

**todo**: tests

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/common/OperationCollection.tsx [hunk 1]
```diff
@@ -69,6 +74,15 @@ const OperatorSelector = styled.div`
...
   flex-direction: column;
+
+  &:after {
+    bottom: -25px;
+    content: '';
+    display: block;
+    height: 5px;
+    position: relative;
+    width: 100%;
+  }
```
Comment (tseho) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/common/OperationCollection.tsx]: This is a hack.
Basically, when used in the modal, the bottom of the box shadow is cropped by the modal container. It's ugly.
I haven't found a way to fix it with padding/margin. This solution add a spacer providing enough space for the box-shadow to be visible.

File: src/Akeneo/Tool/Bundle/MeasureBundle/tests/front/unit/pages/create-measurement-family/hooks/use-create-measurement-family-saver.unit.ts [hunk 2]
```diff
@@ -4,6 +4,14 @@ import '@testing-library/jest-dom/extend-expect';
 import {renderHook} from '@testing-library/react-hooks';
 import {useCreateMeasurementFamilySaver} from 'akeneomeasure/pages/create-measurement-family/hooks/use-create-measurement-family-saver';
 
+declare global {
+  namespace NodeJS {
+    interface Global {
+      fetch: any;
+    }
+  }
+}
```
Comment (tseho) [inline src/Akeneo/Tool/Bundle/MeasureBundle/tests/front/unit/pages/create-measurement-family/hooks/use-create-measurement-family-saver.unit.ts]: We should discuss this, I haven't found a better way yet.

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-measurement-family/form/create-measurement-family-form.ts [hunk 3]
```diff
@@ -3,15 +3,15 @@ import {Operator} from 'akeneomeasure/model/operation';
 
 type LocaleCode = string;
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-measurement-family/form/create-measurement-family-form.ts]: You can now use `model/locale.ts` ;)

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    clearForm,
+    handleClose,
+    setErrors,
+  ]);
+
+  useShortcut(Key.Escape, handleClose);
+
+  return (
+    <Modal>
+      <ModalCloseButton title={__('close')} onClick={handleClose}/>
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx]:
```suggestion
      <ModalCloseButton title={__('pim_common.close')} onClick={handleClose}/>
```

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/form/create-unit-form.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import {Operation, Operator} from 'akeneomeasure/model/operation';
+import {Unit} from 'akeneomeasure/model/unit';
+
+type LocaleCode = string;
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/form/create-unit-form.ts]:
```suggestion
import {LocaleCode} from 'akeneomeasure/model/locale';
```

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/edit/Edit.tsx [hunk 6]
```diff
@@ -53,13 +56,29 @@ const Edit = () => {
   const {measurementFamilyCode} = useParams() as {measurementFamilyCode: string};
   const [currentTab, setCurrentTab] = useState<Tab>(Tab.Units);
   const [measurementFamily, setMeasurementFamily] = useMeasurementFamily(measurementFamilyCode);
+  const [isAddUnitModalOpen, openAddUnitModal, closeAddUnitModal] = useToggleState(false);
+
+  const handleNewUnit = useCallback((unit: Unit) => {
+    null !== measurementFamily && setMeasurementFamily(addUnit(measurementFamily, unit));
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/edit/Edit.tsx]: Maybe you could declare your `handleNewUnit` function after the null check line 68, and then get rid of ` null !== measurementFamily && `

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/components/TextField.tsx [hunk 7]
```diff
@@ -19,29 +25,28 @@ const Input = styled.input.attrs(() => ({className: 'AknTextField'}))<{invalid:
...
   required = false,
-  flag,
   autofocus = false,
+  value,
+  onChange,
+  errors,
+  flag,
   ...props
-}: TextFieldProps & any) => {
+}: TextFieldProps & InputHTMLAttributes<HTMLInputElement>) => {
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/components/TextField.tsx]: 👍

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/icons/TickIcon.tsx [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+
+const TickIcon = ({className = ''}: {className?: string}) => (
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/icons/TickIcon.tsx]: No overridable `title`, `color` & `size`? (as in other Icons)

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+      <ModalBodyWithIllustration illustration={<MeasurementFamilyIllustration/>}>
+        <ModalTitle
+          title={__('measurements.family.add_new_unit')}
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx]: I believe there already is a `measurements.unit.add`

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/translations/jsmessages.en_US.yml [hunk 10]
```diff
@@ -31,6 +31,9 @@ measurements:
             flash:
                 success: Measurement family successfully saved.
                 error: Sorry, an error occured while saving the measurement family.
+        add_new_unit: Add a new unit
+        is_already_used_and_this_new_unit_will_be_read_only_after_creation: A product attribute is linked to this measurement family. Once the unit created, you will only be able to edit its translated labels and symbol.
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/translations/jsmessages.en_US.yml]: That's a pretty long key, isn't something like `measurements.unit.already_used` sufficient?

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/translations/jsmessages.en_US.yml [hunk 11]
```diff
@@ -31,6 +31,9 @@ measurements:
...
+        create_another_unit: Create another unit
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/translations/jsmessages.en_US.yml]: `measurements.unit.create_another`?

File: src/Akeneo/Tool/Bundle/MeasureBundle/tests/front/unit/pages/create-measurement-family/hooks/use-create-measurement-family-saver.unit.ts (hunk 2 above)
Comment (ValentinMumble) [reply src/Akeneo/Tool/Bundle/MeasureBundle/tests/front/unit/pages/create-measurement-family/hooks/use-create-measurement-family-saver.unit.ts]: I'm fine with it, couldn't declare in a test utils file and reuse it?

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  const [isReadOnly, setReadOnly] = useState<boolean>(false);
+  const handleClose = useCallback(() => {
+    clearForm();
+    onClose();
+  }, [clearForm, onClose]);
+  const [errors, setErrors] = useState<ValidationError[]>([]);
+  const measurementFamilyLabel = getMeasurementFamilyLabel(measurementFamily, locale);
+  const measurementFamilyCode = measurementFamily.code;
+
+  // @TODO
```
Comment (juliensnz) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx]: Could you please add an explanatory comment?

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/components/CheckboxField.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  id: string;
+  label: string;
+
+  value: boolean;
+  onChange: (value: boolean) => void;
+  readOnly?: boolean;
+};
+
+const CheckboxField = ({id, label, value, onChange, readOnly}: CheckboxFieldProps) => {
+  return (
```
Comment (juliensnz) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/components/CheckboxField.tsx]: no errors for this field?

File: src/Akeneo/Tool/Bundle/MeasureBundle/tests/front/unit/pages/create-measurement-family/hooks/use-create-measurement-family-saver.unit.ts (hunk 2 above)
Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/tests/front/unit/pages/create-measurement-family/hooks/use-create-measurement-family-saver.unit.ts]: Could be a `.d.ts` file but I don't want to pollute the typescript config with this test-only interface.

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/translations/jsmessages.en_US.yml (hunk 10 above)
Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Resources/translations/jsmessages.en_US.yml]: `measurements.unit.already_used` => `Once the unit is created, you will only be able to edit its translated labels and symbol.` ?
I know mine is too long but I don't want to loose the fact it's really a specific message.
What do you thing about `measurements.unit.will_be_read_only` ?

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx (hunk 4 above)
Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx]: good catch !

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/icons/TickIcon.tsx (hunk 8 above)
Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/icons/TickIcon.tsx]: This is a really really specific icon only used by the checkbox. This is copy-pasted from AM.

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx (hunk 9 above)
Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/pages/create-unit/CreateUnit.tsx]: `Add a new unit` != `Add unit`
I'm following design

File: src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/components/CheckboxField.tsx (hunk 13 above)
Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Resources/public/shared/components/CheckboxField.tsx]: fixed

**Pull request merged**